### PR TITLE
[SPARK-28621][SQL] Make spark.sql.crossJoin.enabled default value true

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

Make `spark.sql.crossJoin.enabled` default value true

### Why are the changes needed?

For implicit cross join, we can set up a watchdog to cancel it if running for a long time.
When "spark.sql.crossJoin.enabled" is false, because `CheckCartesianProducts` is implemented in logical plan stage, it may generate some mismatching error which may confuse end user:
* it's done in logical phase, so we may fail queries that can be executed via broadcast join, which is very fast.
* if we move the check to the physical phase, then a query may success at the beginning, and begin to fail when the table size gets larger (other people insert data to the table). This can be quite confusing.
* the CROSS JOIN syntax doesn't work well if join reorder happens.
* some non-equi-join will generate plan using cartesian product, but `CheckCartesianProducts` do not detect it and raise error.

So that in order to address this in simpler way, we can turn off showing this cross-join error by default.

For reference, I list some cases raising mismatching error here:
Providing:
```
spark.range(2).createOrReplaceTempView("sm1") // can be broadcast
spark.range(50000000).createOrReplaceTempView("bg1") // cannot be broadcast
spark.range(60000000).createOrReplaceTempView("bg2") // cannot be broadcast
```
1) Some join could be convert to broadcast nested loop join, but CheckCartesianProducts raise error. e.g.
```
select sm1.id, bg1.id from bg1 join sm1 where sm1.id < bg1.id
```
2) Some join will run by CartesianJoin but CheckCartesianProducts DO NOT raise error. e.g.
```
select bg1.id, bg2.id from bg1 join bg2 where bg1.id < bg2.id
```

### CODE REVIEW

**Comment by WeichenXu123** [comment]
@cloud-fan @gatorsmile

**Comment by dongjoon-hyun** [comment]
Since this is a reverse of [[SPARK-15425][SQL] Disallow cross joins by default](dafcb05c2ef8e09f45edfb7eabf58116c23975a0), could you update the migration document too? Otherwise, this will surprise administrators.

**Comment by srowen** [comment]
I don't have a strong opinion on this, and probably agree it's better to not fail these by default, is it really just to improve an error message? that seems disproportionate.

**Comment by maropu** [comment]
Yea, I have the same opinion with @srowen; if the current mismatching errors confuse users, we should fix them, first?

**Comment by WeichenXu123** [comment]
@maropu Yep, we can move the `CheckCartesianProducts` to planner stage and fix some mismatching errors.
But note that because there is "join reorder" in optimizer, we have no way to prevent mismatching error in all cases (such as there is a cross join, and then it is reordered...)
So that in order to address this in simpler way, we can turn off showing this cross-join error by default.

**Comment by cloud-fan** [comment]
I do have some complaints about the cross join check:
1. it's done in logical phase, so we may fail queries that can be executed via broadcast join, which is very fast.
2. if we move the check to the physical phase, then a query may success at the beginning, and begin to fail when the table size gets larger (other people insert data to the table). This can be quite confusing.
3. the CROSS JOIN syntax doesn't work well if join reorder happens.

so +1 to disable this check by default.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -733,7 +733,7 @@ object SQLConf {
     .doc("When false, we will throw an error if a query contains a cartesian product without " +
         "explicit CROSS JOIN syntax.")
     .booleanConf
-    .createWithDefault(false)
+    .createWithDefault(true)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: can we also make this config internal? We know that there are problems in the cross join check and we may want to completely get rid of it at some point.

**Comment by maropu** [comment]
We don't need to inform users that their logical plans have cross-joins nodes in `CheckCartesianProducts` via `logInfo` or `logWarn` when `spark.sql.crossJoin.enabled `=true?

**Comment by dilipbiswal** [comment]
I have a question here. So is the plan to fail the query at runtime based on some criteria ? In my thinking the current way was get users to manually opt-in (basically putting the burden on users) as this type of join can be really slow when run on relatively larger tables and may affect the health of the cluster.

**Comment by cloud-fan** [comment]
@dilipbiswal I talked about the problems before, see https://github.com/apache/spark/pull/25520#issuecomment-524328930

**Comment by dilipbiswal** [comment]
@cloud-fan Thanks.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**